test_plans/non_uniform_groups.asciidoc

```diff
@@ -0,0 +1,370 @@
+
+==== get_local_linear_range
+
+Check that `get_local_linear_range()` return type is `linear_id_type` and return
+value is equal to `ballot.get_local_range()` converted to `linear_id_type`.
+
+==== leader
+
+Check that `leader()` return type is `bool` and return value is equal to
+`get_local_id() == 0`.
+
+=== The `fixed_size_group` class API
+
+Let `N` be some power-of-two value greater than 1 that is expected to be a
+divisor of the sub-group size of most devices. The `get_fixed_size_group` is
+called with the `sub_group` of the invocation and `N` as the partition size.
+Let `M` be the size of the sub-group the given `fixed_size_group` was created
+from.
+
+==== Group traits
+
+Check the following:
+
+* `is_group<fixed_size_group<N, sub_group>>::value` is `true`.
+* `is_group_v<fixed_size_group<N, sub_group>>` is `true`.
+* `is_user_constructed_group<fixed_size_group<N, sub_group>>::value` is `true`.
+* `is_user_constructed_group<fixed_size_group<N, sub_group>>` is `true`.
+* `is_fixed_topology_group_v<fixed_size_group<N, sub_group>>::value` is `false`.
+* `is_fixed_topology_group_v<fixed_size_group<N, sub_group>>` is `false`.
+
+==== Members
+
+Check the following:
+
+* `id_type` is same as `id<1>`.
+* `range_type` is same as `range<1>`.
+* `linear_id_type` is same as `uint32_t`.
+* `dimensions` is 1.
+* `fence_scope` is equal to `sub_group::fence_scope`.
+
+==== get_group_id
+
+Check that `get_group_id()` return type is `id<1>` and return value is less than
+`M/N`.
+
+==== get_local_id
+
+Check that `get_local_id()` return type is `id<1>` and return value is less than
+`N`.
+
+==== get_group_range
+
+Check that `get_group_range()` return type is `range_type` and return value is
+equal to `M/N`.
+
+==== get_local_range
+
+Check that `get_local_range()` return type is `range_type` and return value is
+equal to `N`.
+
+==== get_group_linear_id
+
+Check that `get_group_linear_id()` return type is `linear_id_type` and return
+value is equal to `ballot.get_group_id()` converted to `linear_id_type`.
+
+==== get_local_linear_id
+
+Check that `get_local_linear_id()` return type is `linear_id_type` and the
+return value is equal to `ballot.get_local_id()` converted to `linear_id_type`.
+
+==== get_group_linear_range
+
+Check that `get_group_linear_range()` return type is `linear_id_type` and return
+value is equal to `ballot.get_group_range()` converted to `linear_id_type`.
+
+==== get_local_linear_range
+
+Check that `get_local_linear_range()` return type is `linear_id_type` and return
+value is equal to `ballot.get_local_range()` converted to `linear_id_type`.
+
+==== leader
+
+Check that `leader()` return type is `bool` and return value is equal to
+`get_local_id() == 0`.
+
+=== The `tangle_group` class API
+
+The `get_tangle_group` is called with the `sub_group` of the invocation. This
+will only be called by the first `N` items of the sub-group, where `N` is
+strictly less than the size of the sub-group.
```

Comment:
I think it's important for the CTS to deliberately check usage of a tangle in two sides of a branch:
```c++
if (condition) {
  auto tangle = get_tangle_group(sg);
  // checks
}
else {
  auto tangle = get_tangle_group(sg);
  // identical checks
}
```

Compilers are incredibly likely to see code like this and try to be clever by removing the branch, but that ignores the desired `tangle_group` semantics.  If we don't have a test like this, I don't think we can be confident that an implementation actually understands tangles.

Reply:
Currently my WIP tests use something like:

```
parallel_for(... [&](sycl::nd_item<1> it) {
  if (!condition)
    return;
  auto tangle = get_tangle_group(sg);
  // checks
});
```

Do you think it would be fine to separate them, i.e. something like:

```
parallel_for(... [&](sycl::nd_item<1> it) {
  if (!condition)
    return;
  auto tangle = get_tangle_group(sg);
  // checks
});

...

parallel_for(... [&](sycl::nd_item<1> it) {
  if (condition)
    return;
  auto tangle = get_tangle_group(sg);
  // identical checks
});
```

Reply:
No, I don't think this is enough.  The behavior we want to test isn't whether tangles can exist for both `condition` and `!condition`, but rather that they can exist _simultaneously_.  If the tests are split out, there's no risk of a compiler optimizing too aggressively.

It might be clearer if I complete my example.  The concern is that a compiler may see this:
```c++
if (condition) {
  auto tangle = get_tangle_group(sg);
  // checks
}
else {
  auto tangle = get_tangle_group(sg);
  // identical checks
}
```

and replace it with this:
```c++
if (condition || !condition) {
  auto tangle = get_tangle_group(sg);
  // checks
}
```

That transformation is legal for most code, but is illegal for code containing tangle groups.

Reply:
I am of two minds regarding this. On one hand, I agree that an optimizer might elect to make that optimization, but on the other hand I don't think the CTS should make tests based on potential optimizations. An implementation could simply turn off optimizations in the CTS and pass and there could be a plethora of other patterns that could also be used to cause mayhem here, so it seems a little contrived.

Reply:
I see what you mean. Would it be enough to do it just for the API testing? I.e. all the algorithm tests can still just use one control flow? I would assume in general having the two paths in one test should be good enough to weed out the worst cases of this kind of optimizations.

Reply:
Yes, I think that's sufficient.  That will give us some confidence that the groups are being constructed correctly in weird corner cases.  We're already not going to test all possible code they can put in those branches, so I'm okay if we skip the algorithms too!

Reply:
Alright! I have tried specifying this in the test plan.